Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```diff
@@ -0,0 +1,263 @@
+    typealias Element = Result<KafkaAcknowledgedMessage, KafkaAcknowledgedMessageError>
+    /// The producer type used in the system.
+    typealias Producer = NIOAsyncSequenceProducer<Element, NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark, KafkaBackPressurePollingSystem>
+
+    /// The state machine that manages the system's state transitions.
+    let stateMachineLock: NIOLockedValueBox<StateMachine>
+
+    /// Closure that takes care of polling `librdkafka` for new messages.
+    var pollClosure: (() -> Void)?
+    /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+    var sequenceSource: Producer.Source? {
+        get {
+            self.stateMachineLock.withLockedValue { stateMachine in
+                return stateMachine.sequenceSource
+            }
+        }
+        set {
+            self.stateMachineLock.withLockedValue { stateMachine in
+                stateMachine.sequenceSource = newValue
+            }
+        }
+    }
+
+    /// A logger.
+    private let logger: Logger
+
+    /// Initializes the ``KafkaBackPressurePollingSystem``.
+    ///
+    /// - Parameter logger: The logger to be used for logging.
+    init(logger: Logger) {
+        self.stateMachineLock = NIOLockedValueBox(StateMachine())
+        self.logger = logger
+    }
+
+    /// Runs the poll loop with the specified poll interval.
+    ///
+    /// - Parameter pollInterval: The desired time interval between two consecutive polls.
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollInterval: Duration) async {
+        actionLoop: while true {
+            let action = self.stateMachineLock.withLockedValue { $0.nextPollLoopAction() }
+
+            switch action {
+            case .pollAndSleep:
+                self.pollClosure?()
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    self.stateMachineLock.withLockedValue { $0.shutDown() }
+                    break actionLoop
+                }
+            case .suspendPollLoop:
+                await withTaskCancellationHandler {
+                    await withCheckedContinuation { continuation in
+                        self.stateMachineLock.withLockedValue { $0.suspendLoop(continuation: continuation) }
+                    }
+                } onCancel: {
+                    self.stateMachineLock.withLockedValue { $0.shutDown() }
+                }
+            case .shutdownPollLoop:
+                self.stateMachineLock.withLockedValue { $0.shutDown() }
+                break actionLoop
+            }
+        }
+    }
+
+    /// The delivery report callback function that handles acknowledged messages.
+    private(set) lazy var deliveryReportCallback: (UnsafePointer<rd_kafka_message_t>?) -> Void = { messagePointer in
```

Comment:
In the scope of this PR can we make a change so that we don't use `UnsafePointer` here but rather bridge the call-back to something well typed inside the method where we set the callback?